Change search behavior

amkwong opened the pull request:
## Ticket
Partly addresses #43

## Purpose
Add an API to return the best variant given a gene (symbol or gene_id), and rewrite/cleanup searchbox.js using Fetch API

## How to test this
Type different queries into the searchbox (chr:pos, chr:start-end, rsnum, gene_name, garbage input which should fail), and verify that the script is sending you to a page that makes sense:
- chr:pos -> single variant as requested
- chr:start-end -> range as requested
- rsnum -> single variant as requested
- gene_name -> **region view (with padding on both sides) for the tissue with the most significant eQTL** (updated)
- garbage input -> window alert of failure)

## Deployment / configuration notes
(none)

abought commented on pheget/api/__init__.py:
```diff
@@ -2,13 +2,73 @@
 api_blueprint = Blueprint("api", __name__)
 
 
+@api_blueprint.route("/bestvar/<string:symbol>/", methods=["GET"])
```
As a site grows, sometimes it can get hard to remember what a route as called, or the syntax it uses.

One way to keep this organized is to apply common rules throughout, and treat relationships as nested routes. (so everything about a single variant is somewhere under `/variant/`)

Would it make sense to call this something like the following?

`/variant/<string:chrom>_<int:pos>/best/`

If this refers to a gene instead of a variant: we can imagine other gene data in this site in the future. Maybe then the top level entity would be `/gene/symbol/bestvar/`, etc

Note the child URL, and setting consistent syntax for how a variant is defined across both routes.

amkwong replied:
Yes, this refers to a gene (the query is either a gene name or gene_id, and it returns the chr:pos of the variant with the best eQTL signal), so it can be treated as a property of a gene.

`/gene/bestvar/` might work, since it should return the same result if you put in either a short gene name (e.g. "CHRNA5") or the longer gene_id ("ENSG00000169684").

amkwong replied:
I've changed it to /api/<string:symbol>/bestvar/, which I hope makes sense: it returns the "best variant" for a given gene (symbol or ENSG...)

abought commented on pheget/static/js/searchbox.js:
```diff
@@ -1,60 +1,113 @@
+function getBestVar(symbol) {
+    const bestVarURL = `/api/bestvar/${symbol}`;
+    var data = fetch(bestVarURL)
+        .then((response) => response.json())
```
As noted on slack, you'll want to check `resp.ok` before processing the payload. Even a bad request (400) would resolve successfully and be parsed by `.then`. This could lead the UI to hang.

(I always wish there was a more elegant way; let me know if you find one)

https://www.tjvantoll.com/2015/09/13/fetch-and-errors/

amkwong replied:
Error handling added.

abought commented on pheget/static/js/searchbox.js:
```diff
@@ -1,60 +1,113 @@
         const start = cMatchRange[3];
         const end = cMatchRange[4];
-        window.location = `/region/?chrom=${chrom}&start=${start}&end=${end}`;
+        return {type: 'range', chrom: `${chrom}`, start: `${start}`, end: `${end}`};
```
Do we need the string template literals here? (`chrom: ${chrom}`)

Cute syntax trick: if the variable is the same name as an object key, then these two are equivalent:

`{chrom: chrom}` vs `{ chrom }`

amkwong replied:
Thanks for the note on syntax, these have been fixed and changed to Promises. All returns from parseSearchText are promises, and should be handled the same way by parseSearch.

abought commented on pheget/api/__init__.py:
```diff
@@ -2,13 +2,73 @@
+    Given a gene, query an SQL database to find the variant with the strongest association with that gene
+    """
+    # ~UGLY HACK~ If omnisearch did not return a gene_id, then searchbox will fill in "undefined" - immediately abort in that case
+    if symbol == "undefined":
```
If this still necessary, in light of this code?

https://github.com/statgen/pheget/pull/47/files#diff-fb9be9e8a12a6fe4420134c94525c670R73

The basic principle is to avoid making requests you think are unnecessary. (performance wonks like to say "the fastest request is the one not made")

If JS gives a value of `undefined`, we wouldn't ever expect that value to be sent to the server at all.

amkwong replied:
I removed this, it should work fine without.

abought commented on pheget/api/__init__.py:
```diff
@@ -2,13 +2,73 @@
+    if symbol == "undefined":
+        abort(404)
+    # Instead of querying locally, I will use omnisearch to get the relevant parameters (chrom, pos, gene_id)
+    url = f"https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch/?q={symbol}&build=GRCh38"
```
Interesting. Does the backend need to talk to omnisearch, or could this be handled by the frontend? (I am partly asking because this endpoint seems to be triggered in response to an omnisearch call; maybe I am misreading that)

https://github.com/statgen/pheget/pull/47/files#diff-fb9be9e8a12a6fe4420134c94525c670R70-R82

amkwong replied:
This one was a somewhat unfortunate compromise I made in designing the API: 
I wanted it to be usable directly (that is, it will return the best variant given just the name of a gene). Unfortunately, to make the performance of the lookup acceptable requires not just gene_id, but also chrom, start, and end. Since I am not storing the full mapping of gene_ids to chrom:start-end (maybe I shoiuld?), I'm using omnisearch to retrieve those fields.

As you noticed, the searchbox's parsing of the search query _also_ looks up omnisearch, so currently we have the unfortunate double-lookup every time you search a field that isn't chrom:pos, chrom:start-end, or rs#: once to figure out if the query is a gene in searchbox.js, and once to retrieve the chrom/start/end in the API in __init__.

I could change the behavior so that if searchbox hits omnisearch once, it passes chrom/start/end to the API, which would prevent the double-hit.

abought replied:
I'd need to sit and think on this one a bit; nudge me tomorrow? If we can do this without the double request that would be ideal, but it's not the kind of question I should attempt to answer without thought and care.

One of the less obvious ways this can get weird: if the external service is slow to respond it could tie up some server resources in the synchronous python request/ response cycle, which would let a few people hitting our server tie up pheget (if omnisearch is slow to respond, then all our workers would be in use waiting for it). Imagine one person with a full cart of groceries at the front of the "10 items or less" line. Not fun.

There are some deployment changes we can make, like ensuring the use of async workers in gunicorn, that can ensure that one long synchronous request would not prevent others from being serviced. 

But if we can do this in fewer requests, and all in browser, that's certainly ideal. :)

abought commented on pheget/api/__init__.py:
```diff
@@ -2,13 +2,73 @@
+        abort(404)
+    # Instead of querying locally, I will use omnisearch to get the relevant parameters (chrom, pos, gene_id)
+    url = f"https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch/?q={symbol}&build=GRCh38"
+    omniJson = json.load(urllib.request.urlopen(url))["data"][0]
```
If we *are* fetching data on the backend, remember to account for the possibility that the request would fail.

amkwong replied:
My hope is that https://github.com/statgen/pheget/blob/265294173a518672df1a996c26a931892c73f202/pheget/api/__init__.py#L28-L33 will handle any failures.

abought replied:
>  will handle any failures.

Writing distributed applications that depend on an external resource is hard. There are several kinds of failure to be concerned about. This isn't formal terminology, just examples I'm dashing off.

1. Hard failures: the other server is down
2. Soft failures: the other server is slow to respond for reasons outside your control, so this API takes a minute (may or may not ever complete)
3. Deterministic failures: eg the other server receives your request, but returns "400 bad request" in response to that input

For categories 1 & 2, I honestly forget what urllib does (I've generally used `requests`, which wraps some of the edge cases with nicer behaviors and can be told to raise exceptions etc for api errors). For category 3, checking the json payload might be sufficient.

abought commented on pheget/api/__init__.py:
```diff
@@ -2,13 +2,73 @@
+                            "chr" + omniJson["chrom"],
+                            omniJson["start"] - 1000000,
+                            omniJson["end"] + 1000000,
+                            omniJson["gene_id"].split(".")[0] + "._%",
```
It looks like the gene ID is a fuzzy match- maybe ENSG versions?

Is there a situation where we would want to store all the versions in the DB? Or should we standardize on a versioned/unversioned identifier?

amkwong replied:
Yes, I found some differences between our current database and omnisearch's ENSG versions. (I should probably have stripped our own (best variant lookup) database's gene_id's of all version numbers.) My database's version numbers comes directly from the data freeze used by GTEx v8.

My database should never have any gene with multiple version numbers - it's just working off a different data freeze than the current omnisearch. Since GTEx v8 is not going to be updated, I consider the version numbers in our database final. (Omnisearch, on the other hand, can be updated in the future, so keeping the fuzzy match will hopefully help with forward compatibility.)

As for going forward, I could re-generate the best variants lookup database by removing all version numbers from the gene_id's, though I think it's useful to know which version the data is based on. And if we are to keep the version numbers, then fuzzy matching seems like the right way to go.

Do you have any particular concerns about any of these approaches (for compatibility/maintainability/etc reasons)?

abought replied:
> Do you have any particular concerns about any of these approaches (for compatibility/maintainability/etc reasons)?

Wrt the variant identifier mismatch- keeping track of the small differences is indeed a bit annoying. This is a low grade annoyance for a gtex only project, but if we use this on a different eQTL dataset, it's much more severe because data munging would be spread all through the app. Thus using pheget on a different dataset would require tracking down data manipulation in many many places.

In that world, I could imagine normalizing identifiers to be something that happened during an initial intake step, or, at very least, a helper function that makes it clear what is going. We've talked about separating "data" from "views" in other contexts; the "models" module could certainly be beefed up for that purpose.

abought commented on pheget/static/js/searchbox.js:
```diff
@@ -1,60 +1,113 @@
+function getOmniSearch(searchText) {
+    const omniurl = `https://portaldev.sph.umich.edu/api/v1/annotation/omnisearch/?q=${searchText}&build=GRCh38`;
+    var data = fetch(omniurl)
+        .then((response) => response.json())
```
As above, remember to check for `resp.ok` when using fetch.

amkwong replied:
Added error handling for this.

amkwong commented:
The searchbox behaves as intended for me. I added another helper API for when the user explicitly requests a range: the API will return the best tissue and gene within the requested range, which will be used by the region view. Also, I updated the region view html with links to external sources.

I am currently done with this branch (I'll create a new branch for additional new features). I am ready to fix up any problems so that we can merge this into master as soon as possible.

Thanks!

abought commented:
Thanks for the changes. 

I'd still like to take a look at the "double request" issue, and see if there is an alternative. Would it be too long a delay if I waited until Monday to tackle that?

amkwong commented:
Sure, Monday is fine. I'll think about how to fix it too, see if I can get a good solution before Monday.

abought commented:
Thanks for the updates!

I've pushed a few small revisions that 
a) Remove the server-side usage of omnisearch entirely (reflecting the `if gene_id is None` design intent),
b) Added some basic smoketests to the API, and
c) Consolidated a lot of the `.catch` statements. Errors are contextual- there's no global way to handle them because how we message users is a UI issue.

If there are no objections, I will merge by end of day today.

amkwong commented:
I made some minor changes which gives some feedback to the user.

There is one more issue that causes problems: if you search for a gene that exists, but does not have a significant eQTL in our database, then it errors out with "NOT FOUND".  As a test, you can try typing in the gene "BX248415.1". 

I think the default behavior for this kind of search should be to go to the region view of the gene's start/end +/- 500k on the first available tissue ("Adipose_Subcutaneous"). I'll try to rewrite this.

abought commented:
> I think the default behavior for this kind of search should be to go to the region view of the gene's start/end +/- 500k on the first available tissue ("Adipose_Subcutaneous"). I'll try to rewrite this.

Thanks for catching that!

This would be a good example to include in unit tests. It is very hard to build UI if the API involves guesswork to use; unit tests can remove some of the mystery in the future.

amkwong commented:
I added some code to handle the case when the gene exists but no eQTLs are found in our database. Now, when this happens, I run one more API search within the region without specifying the gene, and displays the results (if it finds anything). 
If there are absolutely no signals within +/- 500k of the gene, then it will still fail, but in that case not displaying a region is acceptable behavior.

We can add a test using the gene "AC010507.1", which does not have a strong eQTL and can be found in chr19 near our usual test region.

abought commented:
Good to hear there are ideas for improvements! It doesn't sound like those are blockers, so I will merge this and we can make further tweaks going forward.
